### PR TITLE
Initial support for traffic shaping

### PR DESCRIPTION
**Pull request by johscheuer**
This PR add support for traffic shaping which was implemented by Kubernetes. This allows to limit the network ingress/egress off a pod. More details are in the linked PR.

Fixes: https://github.com/containerd/cri/issues/1150

This PR depends on: https://github.com/containerd/go-cni/pull/40

### CODE REVIEW

File: pkg/server/sandbox_run.go [hunk 1]
```diff
@@ -559,6 +562,16 @@ func (c *criService) setupPod(id string, path string, config *runtime.PodSandbox
 	return "", result, errors.Errorf("failed to find network info for sandbox %q", id)
 }
 
+// toCNIPortMappings converts CRI annotations to CNI bandwidth.
+func toCNIBandWidth(annotations map[string]string) cni.BandWidth {
```
Comment (johscheuer) [inline pkg/server/sandbox_run.go]: We could reuse the functionality from: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/bandwidth/utils.go

what do you mean?

Comment (Random-Liu) [reply pkg/server/sandbox_run.go]: sg

**Comment by johscheuer** [comment]
I just tested my PR with the following pod spec:

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/ingress-bandwidth: 1Gi
    #kubernetes.io/egress-bandwidth: 1M
  name: iperf-server
  labels:
    app: iperf-server
spec:
  restartPolicy: OnFailure
  containers:
  - name: iperf-server
    image: networkstatic/iperf3
    args: [ "-s", "--verbose" ]
    ports:
    - containerPort: 5201
```

after this I ran iperf from a host:

```bash
$ iperf3 -c 192.168.179.130 -p 5201 -i 1 -t 10
Connecting to host 192.168.179.130, port 5201
[  4] local 10.132.0.3 port 56596 connected to 192.168.179.130 port 5201
[ ID] Interval           Transfer     Bandwidth       Retr  Cwnd
[  4]   0.00-1.00   sec   613 MBytes  5.14 Gbits/sec    0    550 KBytes       
[  4]   1.00-2.00   sec   122 MBytes  1.03 Gbits/sec    0    619 KBytes       
[  4]   2.00-3.00   sec   122 MBytes  1.03 Gbits/sec    0    679 KBytes       
[  4]   3.00-4.00   sec   122 MBytes  1.03 Gbits/sec    0    710 KBytes       
[  4]   4.00-5.00   sec   122 MBytes  1.03 Gbits/sec    0    744 KBytes       
[  4]   5.00-6.00   sec   123 MBytes  1.03 Gbits/sec    0    912 KBytes       
[  4]   6.00-7.00   sec   122 MBytes  1.03 Gbits/sec    0    912 KBytes       
[  4]   7.00-8.00   sec   122 MBytes  1.03 Gbits/sec    0    956 KBytes       
[  4]   8.00-9.00   sec   122 MBytes  1.02 Gbits/sec    0    956 KBytes       
[  4]   9.00-10.00  sec   122 MBytes  1.03 Gbits/sec    0    956 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec  1.67 GBytes  1.44 Gbits/sec    0             sender
[  4]   0.00-10.00  sec  1.67 GBytes  1.44 Gbits/sec                  receiver

iperf Done.
```

and the rules are also implemented:

```bash
tc qdisc show
qdisc noqueue 0: dev lo root refcnt 2 
qdisc mq 0: dev ens4 root 
qdisc fq_codel 0: dev ens4 parent :2 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
qdisc fq_codel 0: dev ens4 parent :1 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms memory_limit 32Mb ecn 
qdisc noqueue 0: dev docker0 root refcnt 2 
qdisc noqueue 0: dev tunl0 root refcnt 2 
qdisc noqueue 0: dev cali8ae849af6fa root refcnt 2 
qdisc tbf 1: dev cali7f4c826eaba root refcnt 2 rate 1073Mbit burst 512Mb lat 25.0ms 
qdisc noqueue 0: dev cali84e9fe4a330 root refcnt 2 
```

**Comment by johscheuer** [comment]
@Random-Liu if you have some time the PR is ready for review.

**Comment by Random-Liu** [comment]
`I0528 18:26:33.040] pkg/server/sandbox_run.go:579:25:warning: error strings should not be capitalized (ST1005) (staticcheck)`

**Comment by johscheuer** [comment]
Fixed the issue thanks! Seems like I have overseen it, sorry.

File: pkg/server/sandbox_run.go [hunk 2]
```diff
@@ -559,6 +572,32 @@ func (c *criService) setupPod(id string, path string, config *runtime.PodSandbox
...
+	}
+
+	if ingress != nil {
+		bandWidth.IngressRate = uint64(ingress.Value())
+		bandWidth.IngressBurst = math.MaxUint32
+	}
+
+	if egress != nil {
+		bandWidth.EgressRate = uint64(egress.Value())
+		bandWidth.EgressRate = math.MaxUint32
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: Burst?

File: pkg/server/sandbox_run.go [hunk 3]
```diff
@@ -559,6 +572,32 @@ func (c *criService) setupPod(id string, path string, config *runtime.PodSandbox
...
+// toCNIPortMappings converts CRI annotations to CNI bandwidth.
+func toCNIBandWidth(annotations map[string]string) (*cni.BandWidth, error) {
+	ingress, egress, err := bandwidth.ExtractPodBandwidthResources(annotations)
+	if err != nil {
+		return nil, errors.Errorf("reaing pod bandwidth annotations: %v", err)
+	}
+
+	bandWidth := &cni.BandWidth{}
+
+	if ingress == nil && egress == nil {
```
Comment (Random-Liu) [inline pkg/server/sandbox_run.go]: Not needed.

File: pkg/server/sandbox_run.go (hunk 2 above)
Comment (johscheuer) [reply pkg/server/sandbox_run.go]: Good Catch 👍 Thanks

File: pkg/server/sandbox_run.go (hunk 3 above)
Comment (johscheuer) [reply pkg/server/sandbox_run.go]: Removed it

File: pkg/server/sandbox_run.go [hunk 4]
```diff
@@ -559,6 +572,28 @@ func (c *criService) setupPod(id string, path string, config *runtime.PodSandbox
 	return "", result, errors.Errorf("failed to find network info for sandbox %q", id)
 }
 
+// toCNIPortMappings converts CRI annotations to CNI bandwidth.
```
Comment (jterry75) [inline pkg/server/sandbox_run.go]: NIT: Copy paste error

Comment (jterry75) [reply pkg/server/sandbox_run.go]: There are a few nits. Will send a PR shortly

Comment (jterry75) [reply pkg/server/sandbox_run.go]: Link: https://github.com/containerd/cri/pull/1157

**Pull request merged**